### DISCUSSION TITLE
Getting too many arguments in call to spdy.NewRoundTripperWithProxy error when i try to run my terratest go code which deploys, validates and un-deploys k8s pod to/from AWS EKS

### DISCUSSION THREAD

**Discussion by pandu-bhojaraj-mf** [gruntwork-io/knowledge-base#437] [Help] [answered]
**Problem :**
Getting too many arguments in call to spdy.NewRoundTripperWithProxy error when i try to run my terratest go code which deploys, validates and un-deploys k8s pod to/from AWS EKS

**Problem description** 

Hi All I am getting the error as mentioned in subject when i run my terratest go code. **Actually my scripts were perfectly working fine around 3 months back** but **looks like some library change happened** in between these 3 months in **k8s side which is affecting my scripts** 

Most problematic part is **i am unable to find out at which line number my script is failing** 

go mod init  and go go mod tidy are working fine but as soon as i run go test command getting the error as attached in the screen shot
<img width="960" alt="Issue" src="https://user-images.githubusercontent.com/73541850/170688549-0f2cc749-dc70-4a33-b9b0-a2f731affa75.png">

My code is present in drop box  , here is the link  https://www.dropbox.com/s/qre9zq8jqj9vfmf/ig_data_loader_test.go?dl=0

Please do the **needful as it is blocking me from running my automated tests** 

My go version 

`go version go1.17.8 linux/amd64`

Below are the commands i used for deploying 

`go mod init <folder name under which .go script is present>`
`go mod tidy -compat=1.17`
`go test -v -timeout 120m`

---

<ins datetime="2022-05-27T11:12:44Z">
  <p><a href="https://support.gruntwork.io/hc/requests/108669">Tracked in ticket #108669</a></p>
</ins>

**Comment by zackproser** [comment]
Hello, 
 
It appears that your k8s.io/apimachinery library is out of date. 
 
If you look at the docs for the function you're seeing an error on: https://pkg.go.dev/k8s.io/apimachinery/pkg/util/httpstream/spdy#NewRoundTripperWithProxy you can see that the signature of the NewRoundTripperWithProxy method matches what the error message is telling you. 
 
Meanwhile, your local version of the NewRoundTripperWithProxy method is using a different signature. 
 
This suggests to me that your go.mod and go.sum entries for your [k8s.io/apimachinery](http://k8s.io/apimachinery) library have been updated - but your local source code has not. 
 
To resolve this, I would grep through your local codebase to find every call to NewRoundTripperWithProxy and update those calls to match the updated signature of https://pkg.go.dev/k8s.io/apimachinery/pkg/util/httpstream/spdy#NewRoundTripperWithProxy

Also please note, this is not really a Gruntwork code issue - your question is about how to resolve a Golang dependency issue with Go libraries that Gruntwork does not produce or maintain.

**Reply by pandu-bhojaraj-mf** [reply]
Thanks for your response Zack , I will check this and see how can i apply on my code but one thing is for sure I am not using NewRoundTripperWithProxy function in my code

**Reply by pandu-bhojaraj-mf** [reply]
Hi @zackproser 

**Thanks for you response** , I went through your comments and understood that there is a change in [k8s.io/apimachinery](http://k8s.io/apimachinery) library
But in my script here is the link https://www.dropbox.com/s/cquddgwdo7juhc2/ig-data-loader.go?dl=0 no where i am using signature of https://pkg.go.dev/k8s.io/apimachinery/pkg/util/httpstream/spdy#NewRoundTripperWithProxy function

I am importing [k8s.io/apimachinery](http://k8s.io/apimachinery) library and making use of below 2 functions , in other words calling below 2 functions

[func ListPods(t testing.TestingT, options *KubectlOptions, filters metav1.ListOptions) []corev1.Pod](https://pkg.go.dev/github.com/gruntwork-io/terratest/modules/k8s#ListPods)
[func ListServices(t testing.TestingT, options *KubectlOptions, filters metav1.ListOptions) []corev1.Service](https://pkg.go.dev/github.com/gruntwork-io/terratest/modules/k8s#ListServices)


Here is the code snippet example , **Please let me know if any signature needs to be changed for these two functions** if not please let me know which signature do i need to modify to get rid of this error

import (
metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)

Line # 349 :  pods := k8s.ListPods(t, kubectlOptions, metav1.ListOptions{})
Line # 432 :  services := k8s.ListServices(t, kubectlOptions, metav1.ListOptions{})

**Reply by pandu-bhojaraj-mf** [reply]
Hi @zackproser,

Please help me with this

**Comment by zackproser** [comment]
Hi @pandu-bhojaraj-mf  - if the signature mismatch is happening outside of your own application code, then it's possible some of your dependencies are out of date. 

Looking at your go.mod entry, it appeared that some of the k8s packages being indirectly required were out of date - at v0.20.1 whereas you are trying to use `v0.24.1` of `k8s.io/apimachinery`. 

I was able to get your code to build by running the following commands: 

`go get -u k8s.io/client-go@v0.24.1`
`go mod tidy` 
`go build`

Again, the issue you're facing here is not related in any way to Gruntwork code or applications. You are asking a generic Golang development question.

**Reply by pandu-bhojaraj-mf** [reply]
Hi @zackproser  and all ,

I am able to run this script after running below commands 

`go mod init <script folder name>`
`go get -u k8s.io/client-go@v0.24.1`
`go mod tidy -compat=1.17`
`go test -v -timeout 120m | tee test_output.log`

My go version is as below
$go version
`go version go1.17.8 linux/amd64`

My script is present in below dropbox link
[ig_data_loader_test.go](https://www.dropbox.com/s/qre9zq8jqj9vfmf/ig_data_loader_test.go?dl=0)

Script is being run now but i am getting error when script is trying invoke func [GetNodes](https://github.com/gruntwork-io/terratest/blob/v0.40.17/modules/k8s/node.go#L19) function at line number # 403 `Nodes := k8s.GetNodes(t, kubectlOptions)` 

Here is the screen shot of the error : **exec plugin: invalid apiVersion "client.authentication.k8s.io/v1alpha1"**

<img width="804" alt="image" src="https://user-images.githubusercontent.com/73541850/173833112-74c6397b-e31d-4ba8-b230-e698cb189453.png">

Please **let me know resolution steps** to fix this

**Reply by yorinasub17** [reply]
Hi @pandu-bhojaraj-mf, `client-go` version v0.24.1 syncs with Kubernetes 1.24, which removed the `client.authentication.k8s.io/v1alpha1` API, which you are using in the authentication config in `kubectl`. You want to use `client-go` version `v0.20.6`. Does it work if you do:

```
go get -u k8s.io/client-go@v0.20.6
```

**Reply by yorinasub17** [reply]
Ah I see that this is the original issue you were running into. I suggest doing the following as well:

1. Remove your current `go.sum`
2. In your `go.mod`, make sure the following 3 k8s libraries are locked to the same version as terratest: https://github.com/gruntwork-io/terratest/blob/master/go.mod#L46-L48
3. Run `go mod tidy -compat=1.17`
4. Resolve any compile errors in your current code.

**Reply by pandu-bhojaraj-mf** [reply]
Hi @yorinasub17 

Above steps helped lot actually ,

As part of step #2 `make sure the following 3 k8s libraries are locked to the same version as terratest:` 

I observed difference for _k8s.io/apimachinery_ library in my **go.mod** file

**Actual :** `k8s.io/apimachinery v0.24.1`

I changed to `k8s.io/apimachinery v0.20.6`  manually and tried to run script then **my script ran successfully** after making that manual change 

Now my question is **how can i make sure `k8s.io/apimachinery` is always pointing `v0.20.6` in go.mod file programmatically** as every time i cannot perform this manual change

**Reply by yorinasub17** [reply]
Glad to hear you were able to resolve it! Once you have it in the `go.mod` file, as long as you don't run `go get -u`, it should be locked to that version.

**Comment by pandu-bhojaraj-mf** [comment]
I have posted in stack overflow

https://stackoverflow.com/questions/72521259/getting-too-many-arguments-in-call-to-spdy-newroundtripperwithproxy-error-when-i

**Comment by yorinasub17** [comment, marked as answer]
The k8s library versions need to be in sync with Terratest. You can do the following to ensure they are in sync:

1. Remove your current `go.sum`
2. In your `go.mod`, make sure the following 3 k8s libraries are locked to the same version as terratest: https://github.com/gruntwork-io/terratest/blob/master/go.mod#L46-L48
3. Run `go mod tidy -compat=1.17`
4. Resolve any compile errors in your current code.

**Comment by pandu-bhojaraj-mf** [comment]
Hi @yorinasub17  and @zackproser 

**Thanks a lot** for your help :) . I am able to **run** my script **successfully without any errors** by following below steps  

`go mod init <test script name>`
`go get -u k8s.io/apimachinery@v0.20.6`
`go mod tidy -compat=1.17`
`go test -v -timeout 120m | tee test_output.log`

This discussion can be marked as resolved/**closed** now
